### PR TITLE
stats callback and tests

### PR DESCRIPTION
**Pull request by jonasfj**
Basically adding an optional callback `stats` that will be called with an object...

I'll a utility to `base.stats` that makes reporting this to influxdb trivial. Probably with colums/tags for things like component and process...

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -116,11 +117,33 @@ var makeRequest = function(client, method, url, payload) {
...
+ *
+ * `options.stats` is an optional function that takes an object with the
+ * following properties:
+ * ```js
+ * {
+ *   duration:     0,              // API call time (ms) including all retries
+ *   retries:      0,              // Number of retries
+ *   method:       'createTask',   // Name of method called
+ *   success:      0 || 1,         // Success or error
+ *   resolution:   201,            // Status code
```
Comment (gregarndt) [inline lib/client.js]: Would this be the status code or something like 'http-ok' ? 

Based on: https://github.com/taskcluster/taskcluster-client/pull/20/files#diff-50cfa59973c04321b5da0c6da0fdf4feR289

it looks like the resolution is prefixed with 'http-' and I think res.status is something like 'ok'

File: lib/client.js [hunk 2]
```diff
@@ -277,6 +328,9 @@ exports.createClient = function(reference) {
             }
             debug("Request error calling %s NOT retrying!, err: %s, JSON: %s",
                   entry.name, err, err);
+            if (reportStats) {
+              reportStats(false, (err.code || err.name || 'Error') + '');
```
Comment (gregarndt) [inline lib/client.js]: Why the empty string string being added at the end?

File: lib/client.js (hunk 1 above)
Comment (gregarndt) [inline lib/client.js]: Ok, looking at the test, it looks like it could be something like 'http-201'

File: lib/client.js [hunk 3]
```diff
@@ -240,6 +285,9 @@ exports.createClient = function(reference) {
             // If request was successful, accept the result
             debug("Success calling: %s, (%s retries)",
                   entry.name, attempts - 1);
+            if (reportStats) {
+              reportStats(true, 'http-' + res.status);
```
Comment (gregarndt) [inline lib/client.js]: Why do we prefix this with 'http-' ?

File: lib/client.js (hunk 1 above)
Comment (jonasfj) [inline lib/client.js]: Fixed docs...

File: lib/client.js (hunk 3 above)
Comment (jonasfj) [inline lib/client.js]: So they are strings... also because we use `err.code`

File: lib/client.js (hunk 2 above)
Comment (jonasfj) [inline lib/client.js]: To ensure conversion to string.. stupid really, but it could be something else...

**Comment by jonasfj** [comment]
Landed this with fixes... thanks for the  quick review.

**Pull request merged**